depot/impl.go

```diff
@@ -0,0 +1,141 @@
+package depot
+
+import (
+	"time"
+
+	"github.com/mongodb/anser/bsonutil"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+	"github.com/square/certstrap/depot"
+	mgo "gopkg.in/mgo.v2"
+)
+
+type DBUser struct {
+	ID   string    `bson:"_id"`
+	Cert string    `bson:"cert"`
+	TTL  time.Time `bson:"ttl"`
+}
+
+var (
+	dbUserIDKey   = bsonutil.MustHaveTag(DBUser{}, "ID")
+	dbUserCertKey = bsonutil.MustHaveTag(DBUser{}, "Cert")
+	dbUserTTLKey  = bsonutil.MustHaveTag(DBUser{}, "TTL")
+)
+
+type mongoCertDepot struct {
+	session        *mgo.Session
+	databaseName   string
+	collectionName string
+	expireAfter    time.Duration
+}
+
+type MongoCertDepotOptions struct {
+	MongoDBURI           string
+	MongoDBDialTimeout   time.Duration
+	MongoDBSocketTimeout time.Duration
+	DatabaseName         string
+	CollectionName       string
+	ExpireAfter          time.Duration
+}
+
+// Create a new cert depot in the specified MongoDB.
+func NewMongoCertDepot(opts MongoCertDepotOptions) (depot.Depot, error) {
+	if opts.MongoDBURI == "" {
+		return nil, errors.New("must specify a MongoDB URI!")
+	}
+	if opts.MongoDBDialTimeout <= 0 {
+		opts.MongoDBDialTimeout = 2 * time.Second
+	}
+	if opts.MongoDBSocketTimeout <= 0 {
+		opts.MongoDBSocketTimeout = time.Minute
+	}
+	if opts.DatabaseName == "" {
+		opts.DatabaseName = "certDepot"
+	}
+	if opts.CollectionName == "" {
+		opts.CollectionName = "certs"
+	}
+	if opts.ExpireAfter <= 0 {
+		opts.ExpireAfter = 30 * 24 * time.Hour
+	}
+
+	session, err := mgo.DialWithTimeout("mongodb://localhost:27017", opts.MongoDBDialTimeout)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not connect to db %s", opts.MongoDBURI)
+	}
+	session.SetSocketTimeout(opts.MongoDBSocketTimeout)
+
+	return &mongoCertDepot{
+		session:        session,
+		databaseName:   opts.DatabaseName,
+		collectionName: opts.CollectionName,
+		expireAfter:    opts.ExpireAfter,
+	}, nil
+}
+
+// Put inserts the data into the document specified by the tag.
+func (m *mongoCertDepot) Put(tag *depot.Tag, data []byte) error {
+	if data == nil {
+		return errors.New("data is nil")
+	}
+
+	name := depot.GetNameFromCrtTag(tag)
+	u := &DBUser{
+		ID:   name,
+		Cert: string(data),
+		TTL:  time.Now(),
+	}
+	changeInfo, err := m.session.DB(m.databaseName).C(m.collectionName).UpsertId(name, u)
+	grip.DebugWhen(err == nil, message.Fields{
+		"db":     m.databaseName,
+		"coll":   m.collectionName,
+		"id":     name,
+		"change": changeInfo,
+		"op":     "put",
+	})
+	return errors.Wrap(err, "problem adding data to the database")
+}
+
+// Check returns whether the id/cert pair specified by the tag exists.
+func (m *mongoCertDepot) Check(tag *depot.Tag) bool {
+	name := depot.GetNameFromCrtTag(tag)
+
+	u := &DBUser{}
+	err := m.session.DB(m.databaseName).C(m.collectionName).FindId(name).One(u)
+	grip.WarningWhenf(err != nil && err != mgo.ErrNotFound, "problem looking up '%s' in the database", name)
```

Comment:
assign the conditional to a variable? or have a function? 

also don't we want to log the error value?

Reply:
probably should log more metadata/make this a fields